src/hypothesis/strategies.py

```diff
@@ -1279,6 +1389,42 @@ def __not_a_first_class_strategy(self, name):
             ) % (name,))
     return DataStrategy()
 
+
+def register_type_strategy(custom_type, strategy):
+    """Add an entry to the global type-to-strategy lookup.
+
+    This lookup is used in :func:`~hypothesis.strategies.builds` and
+    :func:`@given <hypothesis.given>`.
+
+    :func:`~hypothesis.strategies.builds` will be used automatically for
+    classes with type annotations on ``__init__`` , so you only need to
+    register a strategy if one or more arguments need to be more tightly
+    defined than their type-based default, or if you want to supply a strategy
+    for an argument with a default value.
+
+    ``strategy`` may be:
+
+    - a search strategy
+    - a function that takes a type and returns a strategy (useful for generic
+      types where unknown subtypes can be handled automatically)
+    - ``None``, in which case ``custom_type`` is removed from the lookup
+
+    """
+    from hypothesis.searchstrategy import types
+    if not isinstance(custom_type, type):
+        raise InvalidArgument('custom_type=%r must be a type')
+    if strategy is None:
+        types._global_type_to_strategy_lookup.pop(custom_type, None)
+    elif not (isinstance(strategy, SearchStrategy) or callable(strategy)):
+        raise InvalidArgument(
+            'strategy=%r must be a SearchStrategy, or a function that takes '
+            'a generic type and returns a specific SearchStrategy')
+    elif isinstance(strategy, SearchStrategy) and strategy.is_empty:
+        raise InvalidArgument('strategy=%r must not be empty')
+    else:
+        types._global_type_to_strategy_lookup[custom_type] = strategy
```

Comment:
If we're *not* going to allow users to override builtins then I think this needs to check whether this is going to be respected and raise an error in the case that it's not. Otherwise people are going to have a debugging nightmare going "I'm defining this so why isn't it appearing in my test!?!?!".